libclamav/udf.c

```diff
@@ -0,0 +1,838 @@
+                        goto done;
+                    }
+                    offset += extDesc->extentLocation.blockNumber * pLogicalVolumeDescriptor->logicalBlockSize;
+                    offset += pPartitionDescriptor->partitionStartingLocation;
+
+                }
+
+                break;
+        default:
+                //impossible unless the file is malformed.
+                cli_warnmsg("extractFile: Unknown descriptor type found.\n");
+                goto done;
+    }
+
+    contents = (uint8_t *)fmap_need_off(ctx->fmap, offset, length);
+    if (NULL == contents) {
+        cli_warnmsg("extractFile: Unable to get offset referenced in the file.\n");
+        goto done;
+    }
+
+    ret = writeWholeFile(ctx, "Test", contents, length);
+
+    fmap_unneed_off(ctx->fmap, offset, length);
+
+done:
+
+    return ret;
+}
+
+static int parseFileEntryDescriptor(cli_ctx *ctx, const uint8_t *const data, PartitionDescriptor *pPartitionDescriptor, LogicalVolumeDescriptor *pLogicalVolumeDescriptor, FileIdentifierDescriptor *fileIdentifierDescriptor)
+{
+
+    FileEntryDescriptor *fed = (FileEntryDescriptor *)data;
+    int ret                  = -1;
+
+    if (261 != fed->tag.tagId) {
+        cli_warnmsg("parseFileEntryDescriptor: Tag ID of 0x%x does not match File Entry Descriptor.\n", fed->tag.tagId);
+        goto done;
+    }
+
+    if (257 != fileIdentifierDescriptor->tag.tagId) {
+        cli_warnmsg("parseFileEntryDescriptor: Tag ID of 0x%x does not match File Identifier Descriptor.\n", fed->tag.tagId);
+        goto done;
+    }
+
+    extractFile(ctx, pPartitionDescriptor, pLogicalVolumeDescriptor,
+            (void *)&(data[getFileEntryDescriptorSize(fed) - fed->allocationDescLen]),
+            fed->icbTag.flags, fileIdentifierDescriptor);
+
+    ret = 0;
+done:
+    return ret;
+}
+
+static void dumpTag(DescriptorTag *dt)
+{
+    fprintf(stderr, "TagId = %d (0x%x)\n", dt->tagId, dt->tagId);
+    fprintf(stderr, "Version = %d (0x%x)\n", dt->descriptorVersion, dt->descriptorVersion);
+    fprintf(stderr, "Checksum = %d (0x%x)\n", dt->checksum, dt->checksum);
+    fprintf(stderr, "Serial Number = %d (0x%x)\n", dt->serialNumber, dt->serialNumber);
+
+    fprintf(stderr, "Descriptor CRC = %d (0x%x)\n", dt->descriptorCRC, dt->descriptorCRC);
+    fprintf(stderr, "Descriptor CRC Length = %d (0x%x)\n", dt->descriptorCRCLength, dt->descriptorCRCLength);
+    fprintf(stderr, "Tag Location = %d (0x%x)\n", dt->tagLocation, dt->tagLocation);
+}
+
+typedef struct {
+    uint8_t structType;
+    char standardIdentifier[5];
+    uint8_t structVersion;
+    uint8_t rest[2041];
+} GenericVolumeStructureDescriptor;
+#define NUM_GENERIC_VOLUME_DESCRIPTORS 3
+
+/*If this function fails, idx will not be updated*/
+static int skipEmptyDescriptors(cli_ctx *ctx, size_t *idxp, size_t *lastOffsetp)
+{
+    int ret         = -1;
+    uint8_t *buffer = NULL;
+    size_t idx      = *idxp;
+    bool allzeros   = true;
+    size_t i;
+
+    while (1) {
+
```

Comment:
```suggestion
    while (1) {
```
Suggestion to remove extra newline